File: runtime/typescript/src/I8.ts
import { Codec } from "./Codec.js";
import { FormatReader, FormatWriter } from "./FormatIO.js";

export type I8 = number;

export const codec: Codec<I8> = {
    async read(reader: FormatReader): Promise<I8> {
        return (await reader.readU8() << 24) >> 24;
    },

    write(writer: FormatWriter, value: I8): Promise<void> {
        return writer.writeU8(value & 0xFF);
    },
};

export function fromInteger(n: bigint): I8 {
    return (Number(n) << 24) >> 24;
}

File: runtime/java/src/main/java/dev/argon/verilization/java_runtime/I8.java
package dev.argon.verilization.java_runtime;

import java.io.IOException;

public final class I8 {
    private I8() {}

    public static byte fromInteger(int i) {
        return (byte)i;
    }

    public static final Codec<Byte> codec = new Codec<Byte>() {
        @Override
        public Byte read(FormatReader reader) throws IOException {
            return reader.readByte();
        }

        @Override
        public void write(FormatWriter writer, Byte value) throws IOException {
            writer.writeByte(value);
        }
    };
}
